pkg/kubelet/server/server.go

```diff
@@ -729,9 +731,36 @@ func (s *Server) getContainerLogs(request *restful.Request, response *restful.Re
 		response.WriteError(http.StatusInternalServerError, fmt.Errorf("unable to convert %v into http.Flusher, cannot show logs", reflect.TypeOf(response)))
 		return
 	}
-	fw := flushwriter.Wrap(response.ResponseWriter)
+
+	var (
+		stdout io.Writer
+		stderr io.Writer
+		fw     = flushwriter.Wrap(response.ResponseWriter)
+	)
+	if utilfeature.DefaultFeatureGate.Enabled(features.SplitStdoutAndStderr) {
+		wantedStream := logOptions.Stream
+		// No stream type specified, default to All
+		if wantedStream == nil {
+			allStream := v1.LogStreamTypeAll
+			wantedStream = &allStream
+		}
+		switch *wantedStream {
+		case v1.LogStreamTypeStdout:
+			stdout, stderr = fw, nil
+		case v1.LogStreamTypeStderr:
+			stdout, stderr = nil, fw
+		case v1.LogStreamTypeAll:
+			stdout, stderr = fw, fw
+		default:
+			_ = response.WriteError(http.StatusBadRequest, fmt.Errorf("invalid stream type %q", *logOptions.Stream))
+			return
+		}
+	} else {
+		stdout, stderr = fw, fw
+	}
+
 	response.Header().Set("Transfer-Encoding", "chunked")
```

Comment:
@SergeyKanzhelev BTW do you think we should have kubelet set a special header if `SplitStdoutAndStderr` feature gate is on, so that kubectl could show a warning if users attempt to request a specific log stream which kubelet doesn't support?

Reply:
yes, need to return error. User may not notice the warning and it is better to fail early than return the wrong result

Reply:
done